sql/core/src/main/scala/org/apache/spark/sql/execution/datasources/DataSource.scala

```diff
@@ -734,30 +735,53 @@ object DataSource extends Logging {
    * Checks and returns files in all the paths.
    */
   private[sql] def checkAndGlobPathIfNecessary(
-      paths: Seq[String],
+      pathStrings: Seq[String],
       hadoopConf: Configuration,
       checkEmptyGlobPath: Boolean,
-      checkFilesExist: Boolean): Seq[Path] = {
-    val allGlobPath = paths.flatMap { path =>
-      val hdfsPath = new Path(path)
-      val fs = hdfsPath.getFileSystem(hadoopConf)
-      val qualified = hdfsPath.makeQualified(fs.getUri, fs.getWorkingDirectory)
-      val globPath = SparkHadoopUtil.get.globPathIfNecessary(fs, qualified)
-
-      if (checkEmptyGlobPath && globPath.isEmpty) {
-        throw new AnalysisException(s"Path does not exist: $qualified")
+      checkFilesExist: Boolean,
+      numThreads: Integer = 40): Seq[Path] = {
+    val qualifiedPaths = pathStrings.map { pathString =>
+      val path = new Path(pathString)
+      val fs = path.getFileSystem(hadoopConf)
+      path.makeQualified(fs.getUri, fs.getWorkingDirectory)
+    }
+
+    // Split the paths into glob and non glob paths, because we don't need to do an existence check
+    // for globbed paths.
+    val (globPaths, nonGlobPaths) = qualifiedPaths.partition(SparkHadoopUtil.get.isGlobPath)
+
+    val globbedPaths =
+      try {
+        ThreadUtils.parmap(globPaths, "globPath", numThreads) { globPath =>
+          val fs = globPath.getFileSystem(hadoopConf)
```

Comment:
Can you pull this out of what's parallelized? here and below? I don't think it has to be retrieved individually for each call

Reply:
https://github.com/apache/spark/pull/25899/commits/8cbc28aad9aa34ad0ab0d16609863cbfbea176d7

This probably never happens, but what if Spark is given paths with different filesystems?
i.e. `spark.read().csv("s3://file1", "hdfs://file2")`

Reply:
Makes sense, I'd leave it.

Reply:
@srowen Sorry, just to clarify, should I revert https://github.com/apache/spark/commit/8cbc28aad9aa34ad0ab0d16609863cbfbea176d7?

Reply:
I reverted it here: https://github.com/apache/spark/pull/25899/commits/105235ce1cdbdfffb912baba02d31abd009f44c2